### PR TITLE
Add build/taskStart and taskFinish to the exception reporting BSP mechanism

### PR DESCRIPTION
**Pull request by MaciejG604**
This resolves the issue with only one $ivy or $dep import error being visible at a time. The behaviour has been checked with metals.

### CODE REVIEW

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 1]
```diff
@@ -287,10 +293,37 @@ final class BspImpl(
 
     preBuild.thenCompose {
       case Left((ex, scope)) =>
+        val targetId  = currentBloopSession.bspServer.targetIds.head
+        val targetUri = targetId.getUri
+        val target    = targetUri.substring(targetUri.lastIndexOf('=') + 1, targetUri.length)
+
+        val taskStartParams = new TaskStartParams(new TaskId("1"))
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala]: Maybe just use:
```
val taskId = new TaskId(UUID.randomUUID().toString)
```

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 2]
```diff
@@ -287,10 +293,37 @@ final class BspImpl(
 
     preBuild.thenCompose {
       case Left((ex, scope)) =>
+        val targetId  = currentBloopSession.bspServer.targetIds.head
+        val targetUri = targetId.getUri
+        val target    = targetUri.substring(targetUri.lastIndexOf('=') + 1, targetUri.length)
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala]: This won't work if targetUri contains `=` twice.

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 3]
```diff
@@ -287,10 +293,37 @@ final class BspImpl(
 
     preBuild.thenCompose {
       case Left((ex, scope)) =>
+        val targetId  = currentBloopSession.bspServer.targetIds.head
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala]: We need to handle the case when `buildTargetId` doesn't exist and when there are several `targetIds`:
```suggestion
        for (targetId <- currentBloopSession.bspServer.targetIds) {
          ...
        }
```

File: modules/build/src/main/scala/scala/build/bsp/BspClient.scala [hunk 4]
```diff
@@ -155,12 +155,24 @@ class BspClient(
...
-        val touchedFiles =
-          allExceptions.flatMap(reportDiagnosticForFiles(targetId, reset = reset)).toSet
+        val touchedFiles = (ex match {
+          case c: CompositeBuildException =>
+            if reset then // send one build exception with reset == true for every file path
+              c.exceptions.groupBy {
+                _.positions.headOption match
+                  case Some(File(Right(path), _, _)) => Some(path)
+                  case _                             => None
+              }
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspClient.scala]: I think you should process all `positions` from list instead only the first.

**Review by lwronski — COMMENTED** [review]
Add some test to `BspTestDefinitions` to check your changes.

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala (hunk 1 above)
Comment (Gedochao) [reply modules/build/src/main/scala/scala/build/bsp/BspImpl.scala]: +1, `UUID`s should do the trick here

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package scala.build.bsp
 
+import ch.epfl.scala.bsp4j._
```
Comment (Gedochao) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala]: don't import `bsp4j._` here.
instead, use the types from there with the `b` prefix imported on the line below.
For example, `b.TaskId`.

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 6]
```diff
@@ -287,10 +293,43 @@ final class BspImpl(
...
-          ex
-        )
+        val taskId = new b.TaskId(UUID.randomUUID().toString)
+
+        for targetId <- currentBloopSession.bspServer.targetScopeIdOpt(scope) do {
+          val idPrefix = "?id="
+          val target = targetId.getUri.substring(
+            targetId.getUri.lastIndexOfSlice(idPrefix) + idPrefix.length,
+            targetId.getUri.length
+          )
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala]:
```suggestion
          val target = targetId.getUri match {
            case s"$_?id=$targetId" => targetId
            case targetIdUri => targetIdUri
          }
```

File: modules/build/src/main/scala/scala/build/bsp/BspClient.scala [hunk 7]
```diff
@@ -182,7 +181,33 @@ class BspClient(
...
+      diags.filter(_.positions.nonEmpty)
+        .flatMap { diag =>
+          diag.positions.map { position =>
+            Diagnostic(diag.message, diag.severity, Seq(position), diag.textEdit)
+          }
+        }
+        .groupBy(_.positions.head match
+          case File(Right(path), _, _) => Some(path)
+          case _                       => None
+        )
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspClient.scala]:
```suggestion
      diags.
        .flatMap { diag =>
          diag.positions.map { position =>
            Diagnostic(diag.message, diag.severity, Seq(position), diag.textEdit)
          }
        }
        .groupBy(_.positions.headOption match
          case Some(File(Right(path), _, _)) => Some(path)
          case _                       => None
        )
```

File: modules/build/src/main/scala/scala/build/bsp/BspClient.scala [hunk 8]
```diff
@@ -182,7 +181,33 @@ class BspClient(
...
+        .filter(_._1.isDefined)
+        .values
+        .toSeq
+        .flatMap { diagsForPath =>
+          reportDiagnosticForFiles(targetId, reset = reset)(diagsForPath.head) ++
+            diagsForPath.tail.flatMap(reportDiagnosticForFiles(targetId))
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspClient.scala]:
```suggestion
            diagsForPath match {
               case head :: tail =>
                  reportDiagnosticForFiles(targetId, reset = reset)(head)
                    ++ tail.flatMap(reportDiagnosticForFiles(targetId))
               case _ => Nil
```

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 9]
```diff
@@ -287,10 +293,43 @@ final class BspImpl(
...
+            Some(targetId),
+            ex
+          )
+
+          val taskFinishParams = new b.TaskFinishParams(taskId, b.StatusCode.ERROR)
+          taskFinishParams.setEventTime(System.currentTimeMillis())
+          taskFinishParams.setMessage(s"Preprocessed '$target'")
+          taskFinishParams.setDataKind(b.TaskDataKind.COMPILE_REPORT)
+
+          val errorNumber = ex match {
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala]:
```suggestion
          val errorSize = ex match {
```

**Review by lwronski — APPROVED** [review]
LGTM

**Review by Gedochao — APPROVED** [review]

**Pull request merged**